src/chart/funnel/funnelLayout.ts

```diff
@@ -324,68 +377,242 @@ export default function funnelLayout(ecModel: GlobalModel, api: ExtensionAPI) {
+            ): void {
+                if (dynamicHeight) {
+                    const start = getLinePointsBySize(pos, resSize / maxSize * viewSize);
+                    index === indices.length - 1 && exitWidth === 100
+                        || (
+                            resSize += sort === 'ascending' ? pieceHeight : -pieceHeight
+                        );
+                    const end = getLinePointsBySize(pos + pieceHeight, resSize / maxSize * viewSize);
+
+                    data.setItemLayout(idx, {
+                        points: start.concat(end.slice().reverse())
+                    });
+                    return;
                 }
-                else {
-                    height = parsePercent(height, viewHeight);
-                    if (sort === 'ascending') {
-                        height = -height;
-                    }
+                else if (dynamicArea) {
+                    const start = getLinePointsBySize(pos, pieceAreaTop);
+                    const end = getLinePointsBySize(pos + pieceHeight, pieceAreaBottom);
+
+                    data.setItemLayout(idx, {
+                        points: start.concat(end.slice().reverse())
+                    });
+                    return;
                 }
+                else if (showRate) {
+                    // data piece
+                    const dataStart = getLinePoints(idx, pos);
+                    let dataEnd;
+                    const val = data.get(valueDim, idx) as number || 0;
+                    if (exitWidth !== undefined && index === indices.length - 1) {
+                        const itemSize = linearMap(val, [min, max], sizeExtent, true);
+                        const exitSize = itemSize * (exitWidth > 100 ? 100 : exitWidth) / 100;
+                        dataEnd = getLinePointsBySize(pos + pieceHeight / 2, exitSize);
+                    }
+                    else {
+                        dataEnd = getLinePoints(idx, pos + pieceHeight / 2);
+                    }
 
-                const start = getLinePoints(idx, y);
-                const end = getLinePoints(nextIdx, y + height);
+                    // rate piece
+                    const rateStart = getLinePoints(idx, pos + pieceHeight / 2);
+                    const rateEnd = getLinePoints(nextIdx, pos + pieceHeight);
 
-                y += height + gap;
+                    // rate label text about
+                    const nextVal = data.get(valueDim, nextIdx) as number || 0;
+                    let rate: number | string = nextVal / val;
+                    rate = 'Rate ' + (rate * 100).toFixed(0) + '%';
+                    if (index === 0) {
+                        firstVal = val;
+                    }
+                    else if (index === indices.length - 1) {
+                        const lastVal = val;
+                        rate = lastVal / firstVal;
+                        rate = 'Overall rate ' + (rate * 100).toFixed(0) + '%';
```

Comment:
Still the default rate string should not be fixed here. It's not `i18n` friendly and hard to be formatted